### PR TITLE
[TTS] Add audio masking to EasyMagpie

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

Add an optional config to randomly dropout autoregressive information supplied to EasyMagpie backbone and local transformer.

**Collection**: [TTS]

# Usage

Add this to the easy magpie model config

```
  feature_masking:
    _target_: nemo.collections.tts.modules.magpietts_modules.FeatureMasking
    hidden_size: 1536
```

**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/tts/models/easy_magpietts.py [hunk 1]
```diff
@@ -1024,6 +1037,9 @@ def process_batch(
             )
 
         # 5. Prepare audio channel embeddings
+        dropout_audio_conditioning = (
+            (self.feature_masking is not None) and self.training and (not dropout_conditional_input)
+        )
```
Comment (blisc) [inline nemo/collections/tts/models/easy_magpietts.py:1042]: Instead of checking self.training instead of a model, can we do it instead of the module like dropout: https://github.com/pytorch/pytorch/blob/v2.13.0/torch/nn/modules/dropout.py#L35

File: nemo/collections/tts/modules/magpietts_modules.py [hunk 2]
```diff
@@ -784,3 +796,77 @@ def sample_maskgit(
...
+        mask_topk = torch.topk(mask_vals, k=max_len, dim=1, sorted=True).values
+        mask_min_val = torch.gather(mask_topk, index=mask_rank, dim=1)
+        mask = mask_vals >= mask_min_val
+
+        # Set values outside the batch item length back to false
+        mask = mask * len_mask
+
+        return mask
+
+    def forward(self, inputs, mask):
```
Comment (blisc) [inline nemo/collections/tts/modules/magpietts_modules.py:858]: Is there a reason that forward doesn't create the mask, but you instead created a apply_dropout() function instead? Seems to go against pytorch design.

**Review by blisc — CHANGES_REQUESTED** [review]
Let's clean it up for merge to main. I believe we have already fast-merged this into the release runs though.

File: nemo/collections/tts/modules/magpietts_modules.py [hunk 3]
```diff
@@ -784,3 +796,77 @@ def sample_maskgit(
         if use_cfg:
             codes = codes[:actual_batch_size]
         return codes
+
+
+class FeatureMasking(NeuralModule):
```
Comment (blisc) [inline nemo/collections/tts/modules/magpietts_modules.py:801]: Let's add a simple test with a known mask (either by fixing random seed, or by creating a mock mask) and ensure that values are masked.

File: nemo/collections/tts/modules/magpietts_modules.py (hunk 2 above)
Comment (rlangman) [reply nemo/collections/tts/modules/magpietts_modules.py]: The purpose of having this method is described in the function comment, which is for supplying custom mask at inference. But I did not want to add an inference path to utilize that in this PR.

But if I remember correctly, pytorch design is that `forward()` should be the method that is called at training time. So technically I could rename this to something else and make `apply_dropout` the `forward()` method.

File: nemo/collections/tts/modules/magpietts_modules.py (hunk 3 above)
Comment (rlangman) [reply nemo/collections/tts/modules/magpietts_modules.py:802]: Added

File: nemo/collections/tts/models/easy_magpietts.py (hunk 1 above)
Comment (rlangman) [reply nemo/collections/tts/models/easy_magpietts.py]: Changed

**Review by blisc — APPROVED** [review]
